cps/kobo.py

```diff
@@ -0,0 +1,523 @@
+        # token is of the form [b64encoded blob].[b64encoded blob 2]
+        if "." in sync_token_header:
+            return SyncToken(raw_kobo_store_token=sync_token_header)
+
+        try:
+            sync_token_json = json.loads(
+                b64decode(sync_token_header + "=" * (-len(sync_token_header) % 4))
+            )
+            validate(sync_token_json, SyncToken.token_schema)
+            if sync_token_json["version"] < SyncToken.MIN_VERSION:
+                raise ValueError
+
+            data_json = sync_token_json["data"]
+            validate(sync_token_json, SyncToken.data_schema_v1)
+        except (exceptions.ValidationError, ValueError) as e:
+            log.error("Sync token contents do not follow the expected json schema.")
+            return SyncToken()
+
+        raw_kobo_store_token = data_json["raw_kobo_store_token"]
+        try:
+            books_last_modified = datetime.utcfromtimestamp(
+                data_json["books_last_modified"]
+            )
+            books_last_created = datetime.utcfromtimestamp(
+                data_json["books_last_created"]
+            )
+        except TypeError:
+            log.error("SyncToken timestamps don't parse to a datetime.")
+            return SyncToken(raw_kobo_store_token=raw_kobo_store_token)
+
+        return SyncToken(
+            raw_kobo_store_token=raw_kobo_store_token,
+            books_last_created=books_last_created,
+            books_last_modified=books_last_modified,
+        )
+
+    def to_headers(self, headers):
+        headers[SyncToken.SYNC_TOKEN_HEADER] = self.build_sync_token()
+
+    def build_sync_token(self):
+        token = {
+            "version": SyncToken.VERSION,
+            "data": {
+                "raw_kobo_store_token": self.raw_kobo_store_token,
+                "books_last_modified": to_epoch_timestamp(self.books_last_modified),
+                "books_last_created": to_epoch_timestamp(self.books_last_created),
+            },
+        }
+        return b64encode_json(token)
+
+
+@kobo.route("/v1/library/sync")
+@login_required
+@download_required
+def HandleSyncRequest():
+    sync_token = SyncToken.from_headers(request.headers)
+    log.info("Kobo library sync request received.")
+
+    # TODO: Limit the number of books return per sync call, and rely on the sync-continuatation header
+    # instead so that the device triggers another sync.
+
+    new_books_last_modified = sync_token.books_last_modified
+    new_books_last_created = sync_token.books_last_created
+    entitlements = []
+
+    # sqlite gives unexpected results when performing the last_modified comparison without the datetime cast.
+    # It looks like it's treating the db.Books.last_modified field as a string and may fail
+    # the comparison because of the +00:00 suffix.
+    changed_entries = (
+        db.session.query(db.Books)
+        .filter(func.datetime(db.Books.last_modified) > sync_token.books_last_modified)
```

Comment:
I found one additional problem. The initial compare will always return no results, as all books in the database are older than the last sync token. So I replaced the > with a !=

Reply:
Hmm that sounds like an odd fix: won't this send back all books but the most recently modified ones?
When I was playing around with this a while back, sending "NewEntitlements" for already known uuids would cause the device to invalidate previously downloaded books. Are you not seeing the same issue with "ChangedEntitlements" ?

It is also curious that you ended up in a state where this was needed: the first time the device syncs, its sync_token.books_last_modified is set to datetime.min so that this comparison should pass on the first sync.

Reply:
What I did:
Start the first sync, no kepubs in the library -> sync completed, and I guess set the sync date to this date 
Then I debugged the code why no files got downloaded to my reader, or at least the date got set and I found out that the this line of code was hiding all ebooks from the reader. 
I changed the line, started a sync, the epub files list has been send to the kobo reader, but no downloaded (because it's not syncing epubs), and therefore the sync date in the calibre-database wasn't updated.
Afterwards I added one kepub, started the sync it got downloaded and I stopped working.
I think it might be a good idea to add only books with a kepub file to the list of changed books, otherwise the reader always sees a lot of files which it doesn't sync anyway. So we could present an output to the task section of calibre web like: Kobo sync started syncing xx books, ...

Reply:
Hmmm I should probably look into this a bit more.
I've added a filter as you suggested in https://github.com/janeczku/calibre-web/pull/1100/commits/d6a974682406edd366cd972b0823002526ec06dc .
I wonder if adding a new format to a book updates that book's last_modified time? (I hope it does...)

Reply:
Ok I suspect that between d6a9746, eca4895 and d5dedfb this issue is likely fixed.